docs: add infinite scroll blog (#2636)
apify/crawlee: website/blog/2024/08-27-how-to-scrape-infinte-scrolling-pages/index.md

When you look into the folder, you'll see many files, but for now, let’s focus on `main.py` and `routes.py`.

In `main.py`, let's change the target location to the Nike website. Then, just to see how scraping will happen, we'll add `headless = False` to the `PlaywrightCrawler` parameters. Let's also increase the maximum requests per crawl option to 100 to see the power of parallel scraping in Crawlee for Python.

The final code will look like this:

```python
import asyncio

from crawlee.playwright_crawler.playwright_crawler import PlaywrightCrawler

from .routes import router

async def main() -> None:
    """The crawler entry point."""
    crawler = PlaywrightCrawler(
        headless=False,
        request_handler=router,
        max_requests_per_crawl=100,
    )

    await crawler.run(
        [
            'https://nike.com/,
        ]
    )

if __name__ == '__main__':
    asyncio.run(main())

```

Now coming to `routes.py`, let’s remove:

```python
await context.enqueue_links()
```
As we don’t want to scrape the whole website. 

Now, if you run the crawler using the command:

```bash
poetry run python -m nike-crawler
````

> As the cookie dialog is blocking us from crawling more than one shoe, let’s get it out of our way through the `context_manager`.

Reviewer: ```suggestion
As the cookie dialog is blocking us from crawling more than one page's worth of shoes, let’s get it out of our way.
```

We're adding the context manager later, this is just a function call.

> We can handle the cookie dialog by going to Chrome dev tools and looking at the `test_id` to accept cookies, which is the `dialog-accept-button`.

Reviewer: ```suggestion
We can handle the cookie dialog by going to Chrome dev tools and looking at the `test_id` of the "accept cookies" button, which is `dialog-accept-button`.
```

> Now, let’s remove the `context.push_data` call and add the code to accept the dialog in routes.py. The updated code will look like this:

Reviewer: ```suggestion
Now, let’s remove the `context.push_data` call that was left there from the project template and add the code to accept the dialog in routes.py. The updated code will look like this:
```

            for link in shoe_listing_links
            if (url := await link.get_attribute('href'))
        ]
    )

@router.handler('listing')
async def listing_handler(context: PlaywrightCrawlingContext) -> None:
    """Handler for shoe listings."""
``` 

### Extract data from product details

Now that we have all the links to the pages with the title “All Shoes,” the next step is to scrape all the products on each page and the information provided on them.

We'll extract each shoe's URL, title, price, and description. Again, let's go to dev tools and extract each parameter's relevant `test_id`. After scraping each of the parameters, we'll use the `context.push_data` function to add it to the local storage. Now let's add the following code to the `listing_handler` and update it in the `routes.py` file:

```python

@router.handler('listing')
async def listing_handler(context: PlaywrightCrawlingContext) -> None:	async def listing_handler(context: PlaywrightCrawlingContext) -> None:
    """Handler for shoe listings."""	    """Handler for shoe listings."""

    await context.enqueue_links(selector='a.product-card__link-overlay', label='detail')

> 

Reviewer: The snippet seems mixed up

@router.handler('detail')
async def detail_handler(context: PlaywrightCrawlingContext) -> None:
    """Handler for shoe details."""

    title = await context.page.get_by_test_id(
        'product_title',
    ).text_content()

    price = await context.page.get_by_test_id(
        'currentPrice-container',
    ).first.text_content()

    description = await context.page.get_by_test_id(
        'product-description',
    ).text_content()

    await context.push_data(
        {
            'url': context.request.loaded_url,
            'title': title,
            'price': price,
            'description': description,
        }
    )
``` 

### Accept Cookies context manager

Since we're dealing with multiple browser pages with multiple links and we want to do infinite scrolling, we may encounter an accept cookie dialog on each page. This will prevent loading more shoes via infinite scroll.

> We'll need to check for cookies on every page, as each one may be opened with a fresh session (no stored cookies) and we'll get the accept cookie dialog even though we already accepted it.

Reviewer: ```suggestion
We'll need to check for cookies on every page, as each one may be opened with a fresh session (no stored cookies) and we'll get the accept cookie dialog even though we already accepted it in another browser window. However, if we don't get the dialog, we want the request handler to work as usual.
```

> The context manager will solve this problem. So, let's build a context manager:

Reviewer: ```suggestion
To solve this problem, we'll try to deal with the dialog in a parallel task that will run in the background. A context manager is a nice abstraction that will allow us to reuse this logic in all the router handlers. So, let's build a context manager:
```
